ra/ra.go

```diff
@@ -1795,8 +1820,293 @@ func (ra *RegistrationAuthorityImpl) revokeCertificate(ctx context.Context, cert
+	}
+	if status.RevokedReason == ocsp.KeyCompromise {
+		return fmt.Errorf("unable to re-revoke serial %q which is already revoked for keyCompromise", serialString)
+	}
+
+	// The new OCSP response has to be back-dated to the original date.
+	ocspResponse, err := ra.CA.GenerateOCSP(ctx, &capb.GenerateOCSPRequest{
+		Serial:    serialString,
+		IssuerID:  int64(issuerID),
+		Status:    string(core.OCSPStatusRevoked),
+		Reason:    int32(ocsp.KeyCompromise),
+		RevokedAt: status.RevokedDate,
+	})
+	if err != nil {
+		return err
+	}
+
+	_, err = ra.SA.UpdateRevokedCertificate(ctx, &sapb.RevokeCertificateRequest{
+		Serial:   serialString,
+		Reason:   int64(ocsp.KeyCompromise),
+		Date:     status.RevokedDate,
+		Response: ocspResponse.Response,
+	})
+	if err != nil {
+		return err
+	}
+
+	ra.revocationReasonCounter.WithLabelValues(revocation.ReasonToString[ocsp.KeyCompromise]).Inc()
+	return nil
+}
+
+// purgeOCSPCache makes a request to akamai-purger to purge the cache entries
+// for the given certificate.
+// TODO(#5152) make the issuerID argument an issuance.IssuerNameID
+func (ra *RegistrationAuthorityImpl) purgeOCSPCache(ctx context.Context, cert *x509.Certificate, issuerID int64) error {
+	issuer, ok := ra.issuersByNameID[issuance.IssuerNameID(issuerID)]
+	if !ok {
+		// TODO(#5152): Remove this fallback (which only gets used when revoking by
+		// serial, so the issuer ID had to be read from the db).
+		issuer, ok = ra.issuersByID[issuance.IssuerID(issuerID)]
+		if !ok {
+			return fmt.Errorf("%w: unable to identify issuer of cert with serial %q", ErrFailedToPurgeOCSP, core.SerialToString(cert.SerialNumber))
+		}
+	}
+
+	purgeURLs, err := akamai.GeneratePurgeURLs(cert, issuer.Certificate)
+	if err != nil {
+		return fmt.Errorf("%w: %s", ErrFailedToPurgeOCSP, err)
+	}
+
+	_, err = ra.purger.Purge(ctx, &akamaipb.PurgeRequest{Urls: purgeURLs})
+	if err != nil {
+		return fmt.Errorf("%w: %s", ErrFailedToPurgeOCSP, err)
+	}
+
+	return nil
+}
+
+// RevokeCertByApplicant revokes the certificate in question. It allows any
+// revocation reason from (0, 1, 3, 4, 5, 9), because Subscribers are allowed to
+// request any revocation reason for their own certificates. However, if the
+// requesting RegID is an account which has authorizations for all names in the
+// cert but is *not* the original subscriber, it overrides the revocation reason
+// to be 5 (cessationOfOperation), because that code is used to cover instances
+// where "the certificate subscriber no longer owns the domain names in the
+// certificate". It does not add the key to the blocked keys list, even if
+// reason 1 (keyCompromise) is requested, as it does not demonstrate said
+// compromise. It attempts to purge the certificate from the Akamai cache, but
+// it does not hard-fail if doing so is not successful, because the cache will
+// drop the old OCSP response in less than 24 hours anyway.
```

Comment:
This feels like an implementation detail of `ra.purgeOCSPCache()` and it doesn't communicate to the reader that they're need to special-case a sentinel error in the event that a failure does occur.